driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -515,7 +517,14 @@ ListenableFuture<Connection> borrowConnection(
       if (routingKey != null) {
         Metadata metadata = manager.cluster.getMetadata();
         Token t = metadata.newToken(partitioner, routingKey);
-        shardId = host.getShardingInfo().shardId(t);
+        shardId = -1;
+        if (keyspace != null && table != null) {
+          assert t instanceof Token.TokenLong64;
+          shardId = metadata.getShardForTabletToken(keyspace, table, (Token.TokenLong64) t, host);
+        }
```

Comment:
Perhaps add protection here against too-large shardId.

Scenario:
1. we get routing information for a tablet
2. node restarts with fewer shards
3. we look up routing information, get a shard that no longer exists

Reply:
Changed 'this' to min(shardCount, 'this')